krb5: add krb5_check_pac option.

sumit-bose opened the pull request:
A new option krb5_check_pac for the krb5 provider is added to control how
the PAC is validated. Currently only the option 'present' which makes the
PAC mandatory, is available.

Additionally the PAC is checked for consistency and compared with user data
read from LDAP.

The PAC responder will be started automatically if the AD provider is configure
but a new option 'implicit_pac_responder' allows to disable this behavior.

Resolves: https://github.com/SSSD/sssd/issues/5868

sumit-bose commented:
Hi @abbra,

I wonder if you can have a short look at the functionality of the tests in
 - https://github.com/sumit-bose/sssd/blob/require_pac/src/providers/ad/ad_pac_common.c#L68
 - https://github.com/sumit-bose/sssd/blob/require_pac/src/providers/ad/ad_pac_common.c#L166
and check if more can be added?

Since you can override the attribute mapping for the user name and the UPN I'm thinking of add options for `krb5_check_pac` like `skip_strict_account_name_check` and `skip_strict_upn_check` which can be set whenever e.g an alternative user name is used and the related comparison with the PAC would fail?

Would it be useful to add an option which would require that not only the PAC is present but also has PAC_UPN_DNS_FLAG_HAS_SAM_NAME_AND_SID set?

Thanks.

bye,
Sumit

abbra commented:
I think you may want to have a strictest check case where PAC_UPN_DNS_FLAG_HAS_SAM_NAME_AND_SID is set, indeed. I would probably set it by default to 'on' in the case of HAVE_STRUCT_PAC_UPN_DNS_INFO_EX and even use it if attribute remapping is 'on' as well. Attribute remapping would still preserve SID to name mapping, presumably so you should be able to find the entry back in the cache...

sumit-bose commented:
> I think you may want to have a strictest check case where PAC_UPN_DNS_FLAG_HAS_SAM_NAME_AND_SID is set, indeed. I would probably set it by default to 'on' in the case of HAVE_STRUCT_PAC_UPN_DNS_INFO_EX and even use it if attribute remapping is 'on' as well. Attribute remapping would still preserve SID to name mapping, presumably so you should be able to find the entry back in the cache...

Hi,

this would work for id-overrides, but not if `ldap_user_name` is set to something other than `sAMAccountName`. But I will think about letting the AD provider read `sAMAccountName` unconditionally and store it in a suitable attribute.

bye,
Sumit

alexey-tikhonov commented:
Hi @sumit-bose,

could you please rebase it to trigger latest CI checks?

sumit-bose commented:
Hi,

the latest version is just a rebase, I'm still thinking about the option names.

bye,
Sumit

alexey-tikhonov commented on src/external/samba.m4:
```diff
@@ -160,4 +160,19 @@ AC_CHECK_MEMBERS([struct PAC_LOGON_INFO.resource_groups], , ,
+                 [[ #include <ndr.h>
+                    #include <gen_ndr/krb5pac.h>
+                    #include <gen_ndr/krb5pac.h>]])
+AC_CHECK_MEMBERS([struct PAC_UPN_DNS_INFO.ex], ,
```
What sense does this check make?
It looks `PAC_UPN_DNS_INFO` is used unconditionally later on.

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -24,9 +24,186 @@
+        return false;
+    }
+
+    for (c = 0; c < 6; c++) {
```
`sizeof(sid->id_auth)` instead of `6`?

sumit-bose replied:
fixed

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -24,9 +24,186 @@
+    /* upn_dns_info has no information which is present in logon_info, so
+     * nothing to check. */
+    if (upn_dns_info->flags == 0) {
+        return EOK;
```
Debug message (as counterpart of "PAC consistency check successful.\n")?

sumit-bose replied:
There are now more options to fine-tune the checks.

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -24,9 +24,186 @@
+        return EOK;
+    }
+
+    /* The user object does not has userPrincipalName set explicitly and the
```
imo, 'has' -> 'have'

sumit-bose replied:
fixed

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -24,9 +24,186 @@
 #include "util/util.h"
 
+static bool
+compare_sid_with_dom_sid_and_rid(const struct dom_sid *sid,
```
Wrap in `#ifdef HAVE_STRUCT_PAC_UPN_DNS_INFO_EX` to suppress compilation warning ('defined but not used') on platform missing modern Samba libs?

sumit-bose replied:
fixed

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -24,9 +24,186 @@
+check_logon_info_upn_dns_info(struct PAC_LOGON_INFO *logon_info,
+                              struct PAC_UPN_DNS_INFO *upn_dns_info)
+{
+    char *delim;
```
`delim` and both function arguments can be `const`.

sumit-bose replied:
fixed

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -67,17 +244,78 @@ errno_t ad_get_data_from_pac(TALLOC_CTX *mem_ctx,
+            talloc_free(*_logon_info);
+            *_logon_info = NULL;
+            talloc_free(*_upn_dns_info);
+            *_upn_dns_info = NULL;
```
`_attributes_info` and `_requester_sid`?

sumit-bose replied:
I removed both because those buffers should be only present in a TGT and we can only look into service tickets.

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -67,17 +244,78 @@ errno_t ad_get_data_from_pac(TALLOC_CTX *mem_ctx,
+
+    if (_logon_info != NULL && *_logon_info == NULL) {
+        ret = ERR_CHECK_PAC_FAILED;
+        goto done;
```
It feels inconsistent that output args are left intact here, while later on [some of] output args are freed in case of error.

Perhaps worth handling in `done:` section?

sumit-bose replied:
fixed

alexey-tikhonov commented on src/providers/krb5/krb5_child.c:
```diff
@@ -1683,20 +1692,48 @@ static krb5_error_code validate_tgt(struct krb5_req *kr)
+                goto done;
+            }
+            if (kr->cli_opts->check_pac != NULL) {
+                DEBUG(SSSDBG_CRIT_FAILURE,
```
Is backtrace useful here or would `SSSDBG_IMPORTANT_INFO` be enough?

sumit-bose replied:
fixed

alexey-tikhonov commented on src/util/sss_krb5.c:
```diff
@@ -973,12 +980,6 @@ krb5_error_code sss_extract_pac(krb5_context ctx,
         goto done;
     }
 
-    ret = unsetenv("_SSS_LOOPS");
```
Was it here by mistake?

sumit-bose replied:
It was just on an odd place, I moved it to be nearer to the actual send.

alexey-tikhonov replied:
But previously it wasn't restored.

alexey-tikhonov commented on src/providers/krb5/krb5_child_handler.c:
```diff
@@ -422,6 +422,18 @@ errno_t set_extra_args(TALLOC_CTX *mem_ctx, struct krb5_ctx *krb5_ctx,
     }
 
+    if (krb5_ctx->check_pac) {
+        extra_args[c] = talloc_asprintf(extra_args,
```
Shouldn't size of `extra_args` be adjusted?

sumit-bose replied:
fixed

alexey-tikhonov commented:
In general looks good to me.
Please see few comments inline.

alexey-tikhonov commented on src/man/sssd-krb5.5.xml:
```diff
@@ -470,6 +470,33 @@ krb5_map_user = joe:juser,dick:richard
+                            FreeIPA domains, if configured. The following
+                            options are available:
+                            <variablelist>
+                            <varlistentry>
```
Imo, true/false would be ok.
Another option could be "validate".

sumit-bose replied:
Hi,

the latest version now has more options and `pac_present` now just checks for the presence of the PAC.

bye,
Sumit

sumit-bose commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -24,9 +24,186 @@
+                        upn_dns_info->upn_name,
+                        (size_t) (delim - upn_dns_info->upn_name)) != 0) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Name part of UPN [%s] and and account name [%s] "
```
fixed

sumit-bose commented:
Hi,

in the latest version I renamed the option to `pac_check` and moved it to the [pac] section in sssd.conf. The main reason is to have this option only defined once. Having it in the [domain/...] section would be irritating if there are multiple domains configured with multiple different values for the option.

Additionally there are now more values to better fine-tune the checks. I'm still working on adding more tests to `test_ad_common.c` but the main code should be done (assuming I do not find issues when adding more tests).

bye,
Sumit

alexey-tikhonov commented on src/man/sssd.conf.5.xml:
```diff
@@ -2215,6 +2229,75 @@ pam_gssapi_indicators_map = sudo:pkinit, sudo-i:pkinit
+                            Apply additional checks on the PAC of the Kerberos
+                            ticket which is available in Active Directory and
+                            FreeIPA domains, if configured. The following
+                            options are available:
```
I would mention multiple options can be enumerated using ',' delimiter. Or maybe it's enough to specify type as a 'string list'.

alexey-tikhonov commented on src/util/pac_utils.c:
```diff
@@ -0,0 +1,150 @@
+    talloc_free(list);
+#ifndef HAVE_PAC_RESPONDER
+    if (ret == EOK && flags != 0) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
```
Perhaps `SSSDBG_IMPORTANT_INFO` fits better here as backtrace will be hardly useful.

sumit-bose replied:
fixed

alexey-tikhonov commented on src/util/util.h:
```diff
@@ -816,4 +816,21 @@ uint64_t get_start_time(void);
+#define CHECK_PAC_UPN_DNS_INFO_EX_PRESENT (1 << 4)
+
+errno_t get_pac_check_config(TALLOC_CTX *mem_ctx, struct confdb_ctx *cdb,
+                             const char **pac_check_str, uint32_t *pac_check_opts);
```
`pac_check_str` what is the purpose of `pac_check_str`? It seems only `_opts` are used.

sumit-bose replied:
removed

alexey-tikhonov commented on src/providers/ad/ad_pac_common.c:
```diff
@@ -24,9 +24,268 @@
+    /* If upn_dns_info is not available we have nothing to check. */
+    if (upn_dns_info == NULL) {
+        if (pac_check_opts & CHECK_PAC_UPN_DNS_INFO_PRESENT) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
```
I would use something like `SSSDBG_FUNC_DATA` here.
It is just a helper function. Imo, it is up to user of this helper to decide on severity of an error.
In general it makes sense to dump backtrace once per error event (logs will be more readable) - in the "top level" function.

alexey-tikhonov commented:
Thank you for the update. I added a few more comments.

sumit-bose commented on src/util/util.h:
```diff
@@ -816,4 +816,21 @@ uint64_t get_start_time(void);
+errno_t get_pac_check_config(TALLOC_CTX *mem_ctx, struct confdb_ctx *cdb,
+                             const char **pac_check_str, uint32_t *pac_check_opts);
+
+const char *check_check_pac_opt(const char *inp, uint32_t *check_pac_flags);
```
fixed

sumit-bose commented:
Hi,

thank you for the review, the latest version should resolve all your comments and the Coverity warning.

bye,
Sumit

alexey-tikhonov commented:
Hi,

does `sssd_krb5_localauth_plugin.8.xml` really belong to this PR?

alexey-tikhonov commented:
@justin-stephenson , could you please proofread man page updates?

justin-stephenson commented on src/man/sssd_krb5_localauth_plugin.8.xml:
```diff
@@ -0,0 +1,70 @@
+            SSSD handles the local names for users from a remote source and can
+            read the Kerberos user principal name from the remote source as
+	    well. With this information SSSD can easily handle the mappings
+            mentions above even if the local name and the Kerberos principal
```
s/mentions/mentioned/

justin-stephenson commented on src/man/sssd_krb5_localauth_plugin.8.xml:
```diff
@@ -0,0 +1,70 @@
+ }
+</programlisting>
+            automatically in the SSSD's public Kerberos configuration snippet
+            directory. If this directory in included in the local Kerberos
```
s/If this directory in/If this directory is/

alexey-tikhonov commented:
Multihost tests failure needs to be addressed.

alexey-tikhonov commented:
> No, they are still used with `_nss_sss_getpwnam_r()` above to resolve the user by principal.

Ah, sure.

Multihost tests are still failing.

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/6071

* `master`
    * 4c7f730b858c09e092cdecec973bd20af32b46d4 - localauth: improve localauth add man page
    * 9c12e962e0fcd6e74c9cc844e2e98e0b0dce79fa - monitor: add implicit_pac_responder option.
    * 30dbecaa9ef6957b74692bf86e34afcf3cafae70 - ad: enable the PAC responder implicitly for AD provider
    * 6970cb1bfd2e50286956ea311167ac77f78c3ee5 - pac: apply new pac check options
    * a28f8a337b9df61615015c045695fd21e9aab13f - krb5: add krb5_check_pac option
    * e57ab1ea5cb0d10e7b689ef72f5862e125bcb17d - tests: add PAC upn_dns_info test
    * 2d52fffdbda1558407fe51b0ad91c0b5bea7bae9 - ad: make new PAC buffers available
* `sssd-2-7`
    * 0dc42cbaa84f45646f6ae13f6ab97a028145498c - localauth: improve localauth add man page
    * fcc1bd84f7e927e2bca9e098c48bd8984c8e398f - monitor: add implicit_pac_responder option.
    * e7163273592eb493c809654908f3904eac2d9c84 - ad: enable the PAC responder implicitly for AD provider
    * 8e265c766c8b1380074026490a5fa8f59d2c4641 - pac: apply new pac check options
    * 1c90333b0241771e227d03b5cbfdc07d00d7a061 - krb5: add krb5_check_pac option
    * d6354e0a9c78281e81adab48a9e3d3f394979891 - tests: add PAC upn_dns_info test
    * a912c1251f0707f4ab5095ad4559823dbbdf0550 - ad: make new PAC buffers available
